File: engine/session.py
"""
ПРУФ · engine.session

Сессия защиты понимания: 15 минут, 6–15 вопросов, все вопросы собраны из конкретных
строк конкретного решения. Вопрос нельзя переиспользовать в другой сессии: он привязан
к отпечатку кода, а варианты перетасованы детерминированно от sha256(отпечаток + id сессии).

Типы вопросов:
    danger        — какая из трёх правок пройдёт мимо тестов (самая опасная)
    gap           — заметят ли тесты эту правку
    first_failure — какой тест упадёт первым
    consequence   — что именно сломается в поведении
"""

from __future__ import annotations

import hashlib
import random
import time
from typing import Any, Dict, List, Optional

from . import analysis as analysis_mod
from . import mutations as mut

SCHEMA_VERSION = "pruf.session/1"

SESSION_SECONDS = 15 * 60
MIN_QUESTIONS = 6
MAX_QUESTIONS = 15
DEFAULT_QUESTIONS = 13
MAX_PER_MUTATION = 2

NO_TEST_OPTION = "Ни один тест не заметит правку"
CRASH_OPTION = "Прогон упадёт целиком: ошибка или таймаут"

QUESTION_WEIGHTS = {"danger": 1.2, "gap": 1.3, "first_failure": 1.0, "consequence": 1.0}
QUESTION_TITLES = {
    "danger": "Какая правка опасна?",
    "gap": "Заметят ли ваши тесты эту правку?",
    "first_failure": "Какой тест упадёт первым?",
    "consequence": "Что именно сломается?",
}
QUESTION_HINTS = {
    "danger": "Опасна та правка, которая меняет поведение и при этом оставляет тесты зелёными.",
    "gap": "Правка уже применена к вашему коду. Решите головой, не запуская тесты.",
    "first_failure": "Смотрите на то, какой тест касается изменённой строки.",
    "consequence": "Важно не название оператора, а последствие для пользователя.",
}

_LETTERS = "ABCDE"


def _rng(*material: Any) -> random.Random:
    digest = hashlib.sha256("|".join(str(part) for part in material).encode("utf-8")).hexdigest()
    return random.Random(int(digest[:16], 16))


def _short_diff(row: Dict[str, Any]) -> str:
    before = (row.get("before") or "").strip()
    after = (row.get("after") or "").strip()
    return "- %s\n+ %s" % (before, after)


def _finalize(raw: List[Dict[str, Any]], rng: random.Random):
    ordered = raw[:]
    rng.shuffle(ordered)
    options: List[Dict[str, Any]] = []
    correct: List[str] = []
    for index, item in enumerate(ordered[:len(_LETTERS)]):
        option_id = _LETTERS[index]
        options.append({
            "id": option_id,
            "label": item["label"],
            "kind": item.get("kind", "text"),
            "ref": item.get("ref"),
        })
        if item.get("correct"):
            correct.append(option_id)
    return options, correct


def _mutation_payload(row: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "id": row["id"],
        "operator": row["operator"],
        "kind": row["kind"],
        "skill": row["skill"],
        "skill_title": row["skill_title"],
        "line": row["line"],
        "before": row["before"],
        "after": row["after"],
        "diff": row["diff"],
    }


def _question(kind: str, row: Dict[str, Any], prompt: str, options, correct, explanation: str) -> Dict[str, Any]:
    return {
        "id": "Q?",
        "type": kind,
        "title": QUESTION_TITLES[kind],
        "hint": QUESTION_HINTS[kind],
        "prompt": prompt,
        "weight": QUESTION_WEIGHTS[kind],
        "skill": row["skill"],
        "skill_title": row["skill_title"],
        "mutation": _mutation_payload(row),
        "options": options,
        "correct": correct,
        "explanation": explanation,
    }


def build_questions(result: Dict[str, Any], session_id: str, question_limit: int = DEFAULT_QUESTIONS) -> List[Dict[str, Any]]:
    """Собирает вопросы из результата мутационного анализа."""
    question_limit = max(1, min(int(question_limit), MAX_QUESTIONS))
    rows: List[Dict[str, Any]] = result.get("mutations", [])
    fingerprint = result.get("fingerprint", "")
    tests = [test["name"] for test in result.get("tests", [])]
    survived = [row for row in rows if row.get("status") == "survived"]
    killed = [row for row in rows if row.get("status") == "killed"]

    buckets: Dict[str, List[Dict[str, Any]]] = {"danger": [], "gap": [], "first_failure": [], "consequence": []}

    # danger: одна выжившая правка против двух пойманных
    if len(killed) >= 2:
        for row in survived:
            rng = _rng(fingerprint, session_id, "danger", row["id"])
            picks = rng.sample(killed, 2)
            raw = [{"label": _short_diff(row), "kind": "diff", "correct": True, "ref": row["id"]}]
            raw += [{"label": _short_diff(other), "kind": "diff", "ref": other["id"]} for other in picks]
            options, correct = _finalize(raw, rng)
            buckets["danger"].append(_question(
                "danger", row,
                "Какая из трёх правок изменит поведение и при этом не уронит ни один ваш тест?",
                options, correct,
                "Правка %s (%s, строка %d) прошла все тесты. %s Остальные две правки тесты поймали."
                % (row["id"], row["kind"], row["line"], row["consequence"]),
            ))

    # gap: заметят ли тесты
    for row in rows:
        rng = _rng(fingerprint, session_id, "gap", row["id"])
        raw: List[Dict[str, Any]] = []
        if row.get("status") == "survived":
            raw.append({"label": NO_TEST_OPTION, "correct": True})
            for name in tests[:2]:
                raw.append({"label": "Упадёт тест %s" % name})
            if len(raw) < 3:
                raw.append({"label": CRASH_OPTION})
            explanation = "Прогон показал: после правки %s все тесты остались зелёными. %s" % (
                row["id"], row.get("suggestion", ""))
        elif row.get("first_failed"):
            raw.append({"label": "Упадёт тест %s" % row["first_failed"], "correct": True})
            others = [name for name in tests if name != row["first_failed"]][:2]
            for name in others:
                raw.append({"label": "Упадёт тест %s" % name})
            raw.append({"label": NO_TEST_OPTION})
            explanation = "Тест %s ловит эту правку: %s" % (row["first_failed"], row["consequence"])
        else:
            raw.append({"label": CRASH_OPTION, "correct": True})
            raw.append({"label": NO_TEST_OPTION})
            for name in tests[:1]:
                raw.append({"label": "Упадёт только тест %s" % name})
            explanation = "После правки %s прогон не дошƫл до ассертов — код упал или завис." % row["id"]
        options, correct = _finalize(raw, rng)
        buckets["gap"].append(_question(
            "gap", row,
            "Правка ниже уже внесена в ваш код (строка %d). Что покажет прогон ваших тестов?" % row["line"],
            options, correct, explanation,
        ))

    # first_failure: какой тест упадёт первым
    if len(tests) >= 2:
        for row in killed:
            if not row.get("first_failed"):
                continue
            rng = _rng(fingerprint, session_id, "first_failure", row["id"])
            raw = [{"label": row["first_failed"], "correct": True}]
            others = [name for name in tests if name != row["first_failed"]][:2]
            raw += [{"label": name} for name in others]
            raw.append({"label": NO_TEST_OPTION})
            options, correct = _finalize(raw, rng)
            buckets["first_failure"].append(_question(
                "first_failure", row,
                "Правка %s внесена в строку %d. Какой из ваших тестов упадёт первым?" % (row["id"], row["line"]),
                options, correct,
                "Первым упал %s. Всего правку заметили тесты: %s."
                % (row["first_failed"], ", ".join(row.get("killed_by") or [row["first_failed"]])),
            ))

    # consequence: что именно сломается
    all_consequences = sorted({operator.consequence for operator in mut.OPERATOR_LIST})
    for row in rows:
        rng = _rng(fingerprint, session_id, "consequence", row["id"])
        distractors = [text for text in all_consequences if text != row["consequence"]]
        if len(distractors) < 2:
            continue
        picks = rng.sample(distractors, 2)
        raw = [{"label": row["consequence"], "correct": True}] + [{"label": text} for text in picks]
        options, correct = _finalize(raw, rng)
        buckets["consequence"].append(_question(
            "consequence", row,
            "Что именно изменится в поведении программы после правки в строке %d?" % row["line"],
            options, correct,
            "Оператор %s (%s): %s" % (row["operator"], row["kind"], row["consequence"]),
        ))

    order = ["danger", "gap", "first_failure", "consequence"]
    questions: List[Dict[str, Any]] = []
    per_mutation: Dict[str, int] = {}
    index = 0
    while len(questions) < question_limit:
        progressed = False
        for kind in order:
            bucket = buckets[kind]
            if index >= len(bucket):
                continue
            progressed = True
            candidate = bucket[index]
            mutation_id = candidate["mutation"]["id"]
            if per_mutation.get(mutation_id, 0) >= MAX_PER_MUTATION:
                continue
            per_mutation[mutation_id] = per_mutation.get(mutation_id, 0) + 1
            questions.append(candidate)
            if len(questions) >= question_limit:
                break
        index += 1
        if not progressed:
            break

    for number, question in enumerate(questions, start=1):
        question["id"] = "Q%d" % number
    return questions


def new_session_id(code: str, salt: Optional[str] = None) -> str:
    material = "%s|%s|%s" % (mut.code_fingerprint(code), salt or "", time.time())
    return "s-" + hashlib.sha256(material.encode("utf-8")).hexdigest()[:12]


def build(
    code: str,
    tests_src: str,
    session_id: Optional[str] = None,
    limit: int = mut.DEFAULT_LIMIT,
    question_limit: int = DEFAULT_QUESTIONS,
    result: Optional[Dict[str, Any]] = None,
    meta: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Готовит сессию защиты. Если анализ уже сделан, передайте его в result."""
    session_id = session_id or new_session_id(code)
    result = result or analysis_mod.analyze(code, tests_src, limit=limit)
    if not result.get("ok"):
        return {
            "schema": SCHEMA_VERSION,
            "ok": False,
            "id": session_id,
            "error": result.get("error", "анализ не выполнен"),
            "stage": result.get("stage"),
            "analysis": result,
        }

    questions = build_questions(result, session_id, question_limit=question_limit)
    now = time.time()
    return {
        "schema": SCHEMA_VERSION,
        "ok": True,
        "id": session_id,
        "status": "active",
        "created_at": now,
        "started_at": now,
        "finished_at": None,
        "seconds_total": SESSION_SECONDS,
        "fingerprint": result.get("fingerprint"),
        "meta": meta or {},
        "code": code,
        "tests": tests_src,
        "summary": {
            "mutation_total": result.get("mutation_total", 0),
            "killed": result.get("killed", 0),
            "survived": result.get("survived", 0),
            "mutation_score_pct": result.get("mutation_score_pct", 0),
            "verdict": result.get("verdict", {}),
            "lines": result.get("lines", 0),
            "sites": result.get("sites", 0),
        },
        "baseline": result.get("baseline", {}),
        "skills": result.get("skills", []),
        "gaps": result.get("gaps", []),
        "tests_report": result.get("tests", []),
        "questions": questions,
        "answers": {},
    }


def remaining_seconds(session: Dict[str, Any]) -> int:
    started = session.get("started_at") or time.time()
    left = int(session.get("seconds_total", SESSION_SECONDS) - (time.time() - started))
    return max(0, left)


def public_view(session: Dict[str, Any]) -> Dict[str, Any]:
    """Вид для клиента: правильные ответы и разборы не уходят до ответа."""
    if not session.get("ok", True):
        return dict(session)
    answers = session.get("answers", {})
    questions = []
    for question in session.get("questions", []):
        item = {key: value for key, value in question.items() if key not in {"correct", "explanation"}}
        answer = answers.get(question["id"])
        if answer:
            item["answer"] = answer
            item["correct"] = question["correct"]
            item["explanation"] = question["explanation"]
        questions.append(item)
    view = {key: value for key, value in session.items() if key not in {"questions", "tests", "code"}}
    view["questions"] = questions
    view["code"] = session.get("code", "")
    view["tests"] = session.get("tests", "")
    view["answered"] = len(answers)
    view["questions_total"] = len(session.get("questions", []))
    view["remaining_seconds"] = remaining_seconds(session)
    return view


def answer(session: Dict[str, Any], question_id: str, option_id: str, seconds: Optional[float] = None) -> Dict[str, Any]:
    """Ответ на вопрос. Ответить можно один раз — второй попытки нет."""
    found = None
    for question in session.get("questions", []):
        if question["id"] == question_id:
            found = question
            break
    if found is None:
        return {"ok": False, "error": "Вопрос %s не найден в этой сессии" % question_id}

    answers = session.setdefault("answers", {})
    if question_id in answers:
        stored = answers[question_id]
        return {
            "ok": True, "repeat": True, "correct": stored["correct"],
            "correct_options": found["correct"], "explanation": found["explanation"],
            "mutation": found["mutation"],
            "progress": {"answered": len(answers), "total": len(session.get("questions", []))},
            "status": session.get("status"),
        }

    valid = {option["id"] for option in found["options"]}
    if option_id not in valid:
        return {"ok": False, "error": "Вариант %s не относится к вопросу %s" % (option_id, question_id)}

    is_correct = option_id in found["correct"]
    answers[question_id] = {
        "option_id": option_id,
        "correct": is_correct,
        "seconds": round(float(seconds), 1) if seconds is not None else None,
        "at": time.time(),
        "order": len(answers) + 1,
    }

    total = len(session.get("questions", []))
    if len(answers) >= total:
        session["status"] = "finished"
        session["finished_at"] = time.time()

    return {
        "ok": True,
        "correct": is_correct,
        "correct_options": found["correct"],
        "explanation": found["explanation"],
        "mutation": found["mutation"],
        "suggestion": next(
            (gap.get("suggestion") for gap in session.get("gaps", []) if gap.get("id") == found["mutation"]["id"]),
            None,
        ),
        "progress": {"answered": len(answers), "total": total},
        "status": session.get("status"),
    }


def finish(session: Dict[str, Any]) -> Dict[str, Any]:
    if session.get("status") != "finished":
        session["status"] = "finished"
        session["finished_at"] = time.time()
    return session

File: web/js/session.js
/* Сессия защиты: таймер, вопросы из строк решения, протокол понимания. */
(function () {
	"use strict"
	var P = window.PRUF
	var S = { id: null, sess: null, qs: [], idx: 0, picked: null, correct: 0, left: 900, tick: null, qAt: 0, proto: null, busy: false }

	var TYPES = {
		danger: "Опасная правка",
		gap: "Дыра в тестах",
		first_failure: "Первый падающий тест",
		consequence: "Последствие правки",
	}

	function $(id) {
		return document.getElementById(id)
	}
	function show(id, on) {
		var el = $(id)
		if (el) el.classList.toggle("hidden", !on)
	}
	function overlay(id, on) {
		var el = $(id)
		if (el) el.classList.toggle("is-open", !!on)
	}

	/* --------------------------------------------------------------- таймер */
	function paintTimer() {
		var el = $("timer")
		el.textContent = P.clock(Math.max(0, S.left))
		el.classList.toggle("is-warn", S.left <= 300 && S.left > 60)
		el.classList.toggle("is-danger", S.left <= 60)
	}

	function startTimer() {
		stopTimer()
		paintTimer()
		S.tick = setInterval(function () {
			S.left -= 1
			paintTimer()
			if (S.left <= 0) {
				stopTimer()
				$("ovTimeStat").textContent = "Отвечено " + S.idx + " из " + S.qs.length + " вопросов. Протокол считается по фактическим ответам."
				overlay("ovTime", true)
			}
		}, 1000)
	}

	function stopTimer() {
		if (S.tick) clearInterval(S.tick)
		S.tick = null
	}

	/* --------------------------------------------------------------- вопрос */
	function paintStats() {
		var sum = (S.sess && S.sess.summary) || {}
		$("stMut").textContent = sum.mutation_total != null ? String(sum.mutation_total) : "\u2014"
		$("stSurvived").textContent = sum.survived != null ? String(sum.survived) : "\u2014"
		$("stScore").textContent = sum.mutation_score_pct != null ? sum.mutation_score_pct + " %" : "\u2014"
		$("stCorrect").textContent = S.correct + " из " + S.idx
	}

	function paintProgress() {
		var total = S.qs.length || 1
		$("progressBar").style.width = Math.round((S.idx / total) * 100) + "%"
		$("qCounter").textContent = Math.min(S.idx + 1, total) + " / " + total
	}

	function renderQuestion() {
		var q = S.qs[S.idx]
		if (!q) return finish()
		S.picked = null
		S.qAt = Date.now()
		$("qType").textContent = TYPES[q.type] || "Вопрос"
		$("qWeight").textContent = "вес " + q.weight
		$("qTitle").textContent = q.title
		$("qPrompt").textContent = q.prompt
		$("qSkill").textContent = q.skill_title || ""
		$("qHint").textContent = q.hint || ""
		$("qHint").classList.add("hidden")
		$("qDiff").innerHTML = P.renderDiff(q.mutation.diff)
		$("sessCode").innerHTML = P.renderCode(S.sess.code, { hot: q.mutation.line })
		var hot = document.querySelector("#sessCode .code-line.is-hot")
		if (hot && hot.scrollIntoView) hot.scrollIntoView({ block: "center" })

		var out = []
		for (var i = 0; i < q.options.length; i++) {
			var o = q.options[i]
			var body = o.kind === "diff" ? '<pre class="diff mb-0">' + P.renderDiff(o.label) + "</pre>" : "<span>" + P.esc(o.label) + "</span>"
			out.push('<button type="button" class="q-option" data-id="' + P.escAttr(o.id) + '"><span class="key">' + P.esc(o.id) + "</span>" + body + "</button>")
		}
		$("qOptions").innerHTML = out.join("")
		show("qFeedback", false)
		$("btnAnswer").disabled = true
		$("btnAnswer").classList.remove("hidden")
		$("btnNext").classList.add("hidden")
		paintProgress()
		paintStats()
	}

	function pick(id) {
		S.picked = id
		var rows = document.querySelectorAll("#qOptions .q-option")
		for (var i = 0; i < rows.length; i++) rows[i].classList.toggle("is-selected", rows[i].getAttribute("data-id") === id)
		$("btnAnswer").disabled = false
	}

	function answer() {
		if (!S.picked || S.busy) return
		var q = S.qs[S.idx]
		S.busy = true
		$("btnAnswer").disabled = true
		P.api
			.sessionAnswer(S.id, { question_id: q.id, option_id: S.picked, seconds: Math.round((Date.now() - S.qAt) / 1000) })
			.then(function (r) {
				var a = r.answer || {}
				S.sess = r.session || S.sess
				if (a.correct) S.correct += 1
				var right = a.correct_options || []
				var rows = document.querySelectorAll("#qOptions .q-option")
				for (var i = 0; i < rows.length; i++) {
					var id = rows[i].getAttribute("data-id")
					var isRight = right.indexOf(id) >= 0
					rows[i].classList.toggle("is-correct", isRight)
					rows[i].classList.toggle("is-wrong", id === S.picked && !isRight)
					rows[i].disabled = true
				}
				$("fbTitle").textContent = a.correct ? "Верно — эту строку вы контролируете" : "Мимо — правка прошла незамеченной"
				$("fbChip").className = "verdict " + (a.correct ? "control" : "none")
				$("fbChip").textContent = a.correct ? "верно" : "ошибка"
				$("fbText").textContent = a.explanation || ""
				$("fbSuggestion").textContent = a.suggestion || ""
				show("qFeedback", true)
				$("btnAnswer").classList.add("hidden")
				var last = S.idx >= S.qs.length - 1
				$("btnNext").textContent = last ? "Завершить и получить протокол" : "Следующий вопрос"
				$("btnNext").classList.remove("hidden")
				S.idx += 1
				paintProgress()
				paintStats()
			})
			.catch(function (e) {
				P.toast(e.message, "err")
				$("btnAnswer").disabled = false
			})
			.then(function () {
				S.busy = false
			})
	}

	function next() {
		if (S.idx >= S.qs.length) return finish()
		renderQuestion()
	}

	/* ------------------------------------------------------------- протокол */
	function finish() {
		if (S.busy) return
		S.busy = true
		stopTimer()
		overlay("ovTime", false)
		P.api
			.sessionFinish(S.id)
			.then(function (r) {
				renderProtocol(r.protocol || r)
			})
			.catch(function (e) {
				P.toast(e.message, "err")
			})
			.then(function () {
				S.busy = false
			})
	}

	function renderProtocol(p) {
		S.proto = p
		var cand = p.candidate || {}
		var task = p.task || {}
		$("pWho").textContent = cand.name || "Кандидат"
		$("pTask").textContent =
			(task.title || "") + " · " + (task.language || "") + " · " + (task.lines || 0) + " строк · " + (task.sites || 0) + " мест правки" + (cand.vacancy ? " · " + cand.vacancy : "")
		$("pControl").textContent = p.control_pct + " %"
		$("pAnswers").textContent = p.questions.correct + " / " + p.questions.total
		$("pScore").textContent = p.mutation.score_pct + " %"
		$("pDuration").textContent = p.duration_clock
		$("pVerdictChip").className = "verdict " + P.verdictClass(p.verdict.level)
		$("pVerdictChip").textContent = p.passed ? "допуск подтверждён" : "допуск не подтверждён"
		$("pVerdictLabel").textContent = p.verdict.label
		$("pVerdictText").textContent = p.verdict.summary + " " + (p.verdict.tests_note || "")
		$("pReproducible").textContent = p.reproducible || ""
		$("pThreshold").textContent = "Порог допуска — " + p.pass_threshold_pct + " %. Вердикт воспроизводим: тот же код и те же мутации дадут те же цифры."

		var skills = p.skills || [],
			sk = []
		for (var i = 0; i < skills.length; i++) {
			var s = skills[i]
			sk.push(
				"<tr><td>" + P.esc(s.title) + "</td><td>" + s.correct + " / " + s.questions + "</td>" +
					'<td><span class="bar"><i class="' + P.barClass(s.score_pct) + '" style="width: ' + s.score_pct + '%"></i></span> ' + s.score_pct + " %</td>" +
					'<td><span class="verdict ' + P.verdictClass(s.verdict) + '">' + P.esc(s.verdict_label) + "</span></td></tr>"
			)
		}
		$("pSkills").innerHTML = sk.join("")

		var risks = p.risks || [],
			rs = []
		for (var k = 0; k < risks.length; k++) {
			var r = risks[k]
			rs.push(
				'<div class="card card-pad-sm" style="background: var(--surface-3)">' +
					'<div class="tiny u-red mb-8">' + P.esc(r.kind) + " · строка " + r.line + " · " + P.esc(r.mutation) + "</div>" +
					'<div class="small">' + P.esc(r.text) + "</div>" +
					(r.action ? '<div class="tiny dim mt-16">' + P.esc(r.action) + "</div>" : "") +
					"</div>"
			)
		}
		$("pRisks").innerHTML = rs.join("") || '<p class="tiny dim mb-0">Рисков не зафиксировано.</p>'

		var quotes = (p.quotes || []).slice(0, 4),
			qs = []
		for (var q = 0; q < quotes.length; q++) {
			var c = quotes[q]
			qs.push(
				'<div class="quote"><div class="tiny dim mb-8">' + P.esc(c.mutation) + " · строка " + c.line + " · " + P.esc(c.kind) + "</div>" +
					'<pre class="diff mb-0">' + P.renderDiff(c.diff) + "</pre></div>"
			)
		}
		$("pQuotes").innerHTML = qs.join("") || '<p class="tiny dim mb-0">Цитат нет.</p>'

		var tl = p.timeline || [],
			ts = []
		for (var t = 0; t < tl.length; t++) {
			var x = tl[t]
			ts.push(
				'<div class="tl-item' + (x.correct ? "" : " is-bad") + '">' +
					'<span class="tl-time">' + P.esc(x.clock) + "</span>" +
					"<b>" + P.esc(x.title) + "</b>" +
					'<div class="tiny dim">' + P.esc(x.mutation) + " · строка " + x.line + " · " + P.esc(x.result) + " · " + x.seconds + " с</div></div>"
			)
		}
		$("pTimeline").innerHTML = ts.join("")

		$("pPdf").href = "/api/protocol/" + encodeURIComponent(p.id) + ".pdf"
		$("pTxt").href = "/api/protocol/" + encodeURIComponent(p.id) + ".txt"
		P.gauge($("pGauge"), p.control_pct)

		try {
			P.store.save({
				id: p.session_id,
				protocol: p.id,
				candidate: cand.name,
				vacancy: cand.vacancy || "",
				task: task.title,
				control_pct: p.control_pct,
				passed: p.passed,
				created_at: p.created_at,
				status: "finished",
			})
		} catch (e) {}

		show("viewLoading", false)
		show("viewSession", false)
		show("viewProtocol", true)
		overlay("ovStart", false)
		window.scrollTo(0, 0)
	}

	/* ------------------------------------------------------------------ старт */
	function fillOverlay() {
		var meta = S.sess.meta || {}
		$("ovWho").textContent = meta.candidate || "Кандидат"
		$("ovTask").textContent = meta.task || "Своё решение"
		$("ovCount").textContent = String(S.sess.questions_total)
		$("ovFinger").textContent = S.sess.fingerprint
		$("sessWho").textContent = meta.candidate || "Кандидат"
		$("sessTask").textContent = (meta.task || "") + " · " + (meta.language || "Python")
		$("sessFinger").textContent = "отпечаток " + S.sess.fingerprint
	}

	function useSession(s) {
		S.sess = s
		S.id = s.id
		S.qs = s.questions || []
		S.idx = s.answered || 0
		S.left = s.remaining_seconds != null ? s.remaining_seconds : s.seconds_total || 900
		var answers = s.answers || {}
		S.correct = 0
		for (var key in answers) if (answers[key] && answers[key].correct) S.correct += 1
		show("viewLoading", false)
		fillOverlay()
		if (s.status === "finished") {
			var pid = String(s.id).replace(/^s-/, "p-")
			return P.api
				.protocol(pid)
				.then(function (r) {
					renderProtocol(r.protocol || r)
				})
				.catch(function () {
					finish()
				})
		}
		overlay("ovStart", true)
	}

	function begin() {
		overlay("ovStart", false)
		show("viewSession", true)
		renderQuestion()
		startTimer()
	}

	function fail(msg) {
		show("viewLoading", false)
		show("viewError", true)
		if (msg) $("errText").textContent = msg
	}

	function init() {
		$("btnStart").addEventListener("click", begin)
		$("btnAnswer").addEventListener("click", answer)
		$("btnNext").addEventListener("click", next)
		$("btnFinish").addEventListener("click", finish)
		$("btnTimeFinish").addEventListener("click", finish)
		$("btnHint").addEventListener("click", function () {
			$("qHint").classList.toggle("hidden")
		})
		$("qOptions").addEventListener("click", function (e) {
			var row = e.target.closest(".q-option")
			if (row && !row.disabled) pick(row.getAttribute("data-id"))
		})
		$("btnCopyLink").addEventListener("click", function () {
			P.copy(location.origin + "/session.html?id=" + encodeURIComponent(S.id))
		})
		$("btnSend").addEventListener("click", function () {
			var mail = $("sendEmail").value.trim()
			if (!mail || mail.indexOf("@") < 0) {
				P.toast("Укажите email", "err")
				return
			}
			P.api
				.protocolSend(S.proto.id, mail)
				.then(function (r) {
					$("sendStatus").textContent = r.message || "Протокол сохранён в очереди отправки."
					P.toast("Протокол отправлен", "ok")
				})
				.catch(function (e) {
					P.toast(e.message, "err")
				})
		})
		document.addEventListener("visibilitychange", function () {
			if (document.hidden || !S.id || !S.tick) return
			P.api
				.sessionGet(S.id)
				.then(function (s) {
					if (s.remaining_seconds != null) {
						S.left = s.remaining_seconds
						paintTimer()
					}
				})
				.catch(function () {})
		})

		var id = P.qs("id", "")
		var demo = P.qs("demo", "")
		if (id) {
			P.api
				.sessionGet(id)
				.then(useSession)
				.catch(function (e) {
					fail(e.message)
				})
			return
		}
		if (demo) {
			P.api
				.demo("orders")
				.then(function (d) {
					return P.api.sessionStart({
						code: d.code,
						tests: d.tests,
						meta: { candidate: "Демо-кандидат", task: "Разбор заказов из текстового файла", language: "Python", vacancy: "Python-разработчик" },
						limit: 15,
						questions: 13,
					})
				})
				.then(function (s) {
					if (history.replaceState) history.replaceState(null, "", "/session.html?id=" + encodeURIComponent(s.id))
					useSession(s)
				})
				.catch(function (e) {
					fail(e.message)
				})
			return
		}
		fail("Откройте ссылку на сессию или запустите демо-сессию.")
	}

	if (document.readyState === "loading") document.addEventListener("DOMContentLoaded", init)
	else init()
})()
